File: pintia/gplt_practice/l2/_013.cpp
#include <cstdio>
#include <vector>
#include <string>

using namespace std;

const int N = 510;
int n, m, k;
vector<int> conn[N];
int attacks[N];
bool lost[N];
int roots[N];
int group;
vector<string> ans;

int find(int x) {
    return roots[x] == x ? x : (roots[x] = find(roots[x]));
}

void join(int p, int q) {
    p = find(p), q = find(q);
    if (p == q) return;
    roots[p] = q;
    --group;
}

int main() {
    scanf("%d%d", &n, &m);
    while (m-- > 0) {
        int a, b;
        scanf("%d%d", &a, &b);
        conn[a].push_back(b);
        conn[b].push_back(a);
    }
    scanf("%d", &k);
    if (k == n) ans.push_back("Game Over.");
    for (int i = 0; i < k; ++i) {
        scanf("%d", &attacks[i]);
        lost[attacks[i]] = true;
    }
    for (int i = 0; i < n; ++i) roots[i] = i;
    group = n;
    for (int u = 0; u < n; ++u) {
        if (lost[u]) continue;
        for (auto& v : conn[u]) {
            if (!lost[v]) join(u, v);
        }
    }
    int prev = group;
    for (int i = k - 1; i >= 0; --i) {
        int u = attacks[i];
        lost[u] = false;
        for (auto& v : conn[u]) {
            if (!lost[v]) join(u, v);
        }
        if (prev - group < 2) ans.push_back("City " + to_string(u) + " is lost.");
        else ans.push_back("Red Alert: City " + to_string(u) + " is lost!");
        prev = group;
    }
    for (int i = ans.size() - 1; i >= 0; --i) {
        printf("%s\n", ans[i].c_str());
    }

    return 0;
}

File: pintia/gplt_practice/l2/_013.java
import java.io.*;
import java.util.*;

public class _013 {
    static class UnionFind {
        private int[] roots;
        private int group;

        public UnionFind(int size) {
            roots = new int[size];
            group = size;
            for (int i = 0; i < size; i++) roots[i] = i;
        }

        public int find(int x) {
            return roots[x] == x ? x : (roots[x] = find(roots[x]));
        }

        public void union(int p, int q) {
            p = find(p);
            q = find(q);
            if (p == q) return;
            roots[p] = q;
            --group;
        }

        public int getGroupCount() {
            return group;
        }
    }

    public static void main(String[] args) throws IOException {
        StreamTokenizer in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
        in.nextToken(); int n = (int) in.nval;
        in.nextToken(); int m = (int) in.nval;
        List<Integer>[] conn = new List[n];
        for (int i = 0; i < n; ++i) conn[i] = new ArrayList<>();
        for (int i = 0; i < m; ++i) {
            in.nextToken(); int a = (int) in.nval;
            in.nextToken(); int b = (int) in.nval;
            conn[a].add(b);
            conn[b].add(a);
        }
        in.nextToken(); int k = (int) in.nval;
        List<String> ans = new ArrayList<>();
        if (k == n) ans.add("Game Over.\n");
        int[] attacks = new int[k];
        boolean[] lost = new boolean[n];
        for (int i = 0; i < k; i++) {
            in.nextToken(); attacks[i] = (int) in.nval;
            lost[attacks[i]] = true;
        }
        UnionFind uf = new UnionFind(n);
        for (int u = 0; u < n; u++) {
            if (lost[u]) continue;
            for (int v : conn[u]) {
                if (!lost[v]) uf.union(u, v);
            }
        }
        int prev = uf.getGroupCount();
        for (int i = k - 1; i >= 0; i--) {
            int u = attacks[i];
            lost[u] = false;
            for (int v : conn[u]) {
                if (!lost[v]) uf.union(u, v);
            }
            if (prev - uf.getGroupCount() < 2) {
                ans.add("City " + u + " is lost.\n");
            } else {
                ans.add("Red Alert: City " + u + " is lost!\n");
            }
            prev = uf.getGroupCount();
        }
        StringBuilder sb = new StringBuilder();
        for (int i = ans.size() - 1; i >= 0; i--) {
            sb.append(ans.get(i));
        }
        System.out.print(sb);
    }
}
